core/node/consensus/src/batch.rs

```diff
@@ -0,0 +1,266 @@
+        let batch = conn
+            .batch(ctx, n)
+            .await
+            .wrap("batch()")?
+            .context("batch not in storage")?;
+
+        let proofs = tree
+            .get_proofs(
+                n,
+                vec![
+                    Self::current_l2_block_info_key(),
+                    Self::tx_rolling_hash_key(),
+                ],
+            )
+            .await
+            .context("get_proofs()")?;
+        if proofs.len() != 2 {
+            return Err(anyhow::format_err!("proofs.len()!=2").into());
+        }
+        let current_l2_block_info = proofs[0].clone();
+        let tx_rolling_hash = proofs[1].clone();
+        let (block_number, _) = unpack_block_info(current_l2_block_info.value.as_bytes().into());
+        let prev = L2BlockNumber(
+            block_number
+                .checked_sub(1)
+                .context("L2BlockNumber underflow")?
+                .try_into()
+                .context("L2BlockNumber overflow")?,
+        );
+        let proofs = tree
+            .get_proofs(n, vec![Self::l2_block_hash_entry_key(prev)])
+            .await
+            .context("get_proofs()")?;
+        if proofs.len() != 1 {
+            return Err(anyhow::format_err!("proofs.len()!=1").into());
+        }
+        let l2_block_hash_entry = proofs[0].clone();
+        Ok(Self {
+            info: i_executor::structures::StoredBatchInfo::from(&batch),
+            protocol_version: batch
+                .header
+                .protocol_version
+                .context("missing protocol_version")?,
+
+            current_l2_block_info,
+            tx_rolling_hash,
+            l2_block_hash_entry,
+        })
+    }
+
+    /// Verifies the proof against the commit and returns the hash
+    /// of the last L2 block.
+    pub(crate) fn verify(&self, comm: &LastBlockCommit) -> anyhow::Result<(L2BlockNumber, H256)> {
+        // Verify info.
+        anyhow::ensure!(comm.info == self.info.hash());
+
+        // Check the protocol version.
+        anyhow::ensure!(
+            self.protocol_version >= ProtocolVersionId::Version13,
+            "unsupported protocol version"
+        );
+
+        let (block_number, block_timestamp) =
+            unpack_block_info(self.current_l2_block_info.value.as_bytes().into());
+        let prev = L2BlockNumber(
+            block_number
+                .checked_sub(1)
+                .context("L2BlockNumber underflow")?
+                .try_into()
+                .context("L2BlockNumber overflow")?,
+        );
+
+        // Verify merkle paths.
+        self.current_l2_block_info
+            .verify(Self::current_l2_block_info_key(), self.info.batch_hash)?;
+        self.tx_rolling_hash
+            .verify(Self::tx_rolling_hash_key(), self.info.batch_hash)?;
+        self.l2_block_hash_entry
+            .verify(Self::l2_block_hash_entry_key(prev), self.info.batch_hash)?;
```

Comment:
Nit: Maybe add `context` explaining which proof verification failed exactly? Would probably speed up debugging if ever occurs in a real env.

Reply:
done